sdk/go/common/apitype/search.go

```diff
@@ -0,0 +1,120 @@
+	Created      *string                      `json:"created,omitempty" csv:"created"`
+	Custom       *bool                        `json:"custom,omitempty" csv:"custom"`
+	Delete       *bool                        `json:"delete,omitempty" csv:"delete"`
+	Dependencies []string                     `json:"dependencies,omitempty" csv:"-"`
+	ID           *string                      `json:"id,omitempty" csv:"id"`
+	Matches      map[string][]json.RawMessage `json:"matches,omitempty" csv:"-"`
+	Modified     *string                      `json:"modified,omitempty" csv:"modified"`
+	Module       *string                      `json:"module" csv:"module"`
+	Name         *string                      `json:"name,omitempty" csv:"name"`
+	Package      *string                      `json:"package" csv:"package"`
+	Parent       *string                      `json:"parent.urn,omitempty" csv:"parent_urn"`
+	Pending      *string                      `json:"pending,omitempty" csv:"pending"`
+	Program      *string                      `json:"project,omitempty" csv:"project"`
+	Protected    *bool                        `json:"protected,omitempty" csv:"protected"`
+	ProviderURN  *string                      `json:"provider.urn,omitempty" csv:"provider_urn"`
+	Stack        *string                      `json:"stack,omitempty" csv:"stack"`
+	Type         *string                      `json:"type,omitempty" csv:"type"`
+	URN          *string                      `json:"urn,omitempty" csv:"urn"`
+	Teams        teamNameSlice                `json:"teams,omitempty" csv:"teams"`
+	Properties   *RawProperty                 `json:"properties,omitempty" csv:"properties"`
+}
+
+// Aggregation collects the top 5 aggregated values for the requested dimension.
+type Aggregation struct {
+	Others  *int64              `json:"others,omitempty"`
+	Results []AggregationBucket `json:"results,omitempty"`
+}
+
+// AggregationBucket is an example value for the requested aggregation, with a
+// count of how many resources share that value.
+type AggregationBucket struct {
+	Name  *string `json:"name,omitempty"`
+	Count *int64  `json:"count,omitempty"`
+}
+
+// ResourceSearchPagination provides links for paging through results.
+type ResourceSearchPagination struct {
+	Previous *string `json:"previous,omitempty"`
+	Next     *string `json:"next,omitempty"`
+	Cursor   *string `json:"cursor,omitempty"`
+}
+
+type teamNameSlice []string
+
+func (teams *teamNameSlice) MarshalCSV() (string, error) {
+	if teams == nil || len(*teams) == 0 {
+		return "", nil
+	}
+
+	json, err := json.Marshal(*teams)
+	if err != nil {
+		return "", err
+	}
+
+	return string(json), nil
+}
+
+// RawProperty wraps a json.RawProperty for JSON or CSV export.
+type RawProperty struct {
+	json.RawMessage
+}
+
+type PulumiQueryResponse struct {
+	Query string `json:"query"`
+}
+
+type PulumiQueryRequest struct {
+	Query string `url:"query"`
+}
+
+func ParseQueryParams(rawParams *[]string) *PulumiQueryRequest {
```

Comment:
Slices are almost never passed by pointer. This should be a `[]string`.

Reply:
Why is that? Convention? I'm more than happy to pass by value where possible, but curious what the motivation is here.

Reply:
Slices are already reference types (kind-of; they're triples of pointer, length, and capacity, but they're nilable so for our purposes, they're reference types[^1]), so they're always used and passed by value.
APIs in Go almost never deal with pointers to slices—except in odd cases when you want to, say, modify the slice variable someone else owns. This is similar to using, say `ok = combine(x, y, &result)` in C—to get output from a function into a pointer—and not a thing that's normally done otherwise.

So, less convention and more idiom/correctness. Slices, maps, and interfaces are always passed by value because that's the only way you can consume them. 

[^1]: If you're curious about the internals, https://go.dev/blog/slices-intro has a run-down. I also have a blog post [here](https://abhinavg.net/2019/07/11/zero-alloc-slice-filter/) that tries to explain it with a different example.